File: Ahorn/entities/ParticleEmitter.jl
module FemtoHelperParticleEmitter

using ..Ahorn, Maple

@pardef ParticleEmitter(
x1::Integer,
y1::Integer,
x2::Integer=x1 + 16,
y2::Integer=y1,
particleColor::String="ff7777",
particleColor2::String="9c2e36",
particleTexture::String="particles/feather",
particleFadeMode::Integer=2,
particleColorMode::Integer=3,
particleRotationMode::Integer=1,
particleLifespanMin::Number=1.25,
particleLifespanMax::Number=1.75,
particleSize::Number=1.5,
particleSizeRange::Number=0.5,
particleScaleOut::Bool=true,
particleSpeedMin::Number=10.0,
particleSpeedMax::Number=20.0,
particleFriction::Number=50.0,
particleAccelX::Number=30.0,
particleAccelY::Number=-60.0,
particleFlipChance::Bool=true,
particleSpinSpeedMin::Number=4.0,
particleSpinSpeedMax::Number=8.0,
particleCount::Integer=1,
particleSpawnSpread::Number=4,
spawnInterval::Number=0.1,
spawnChance::Number=90.0,
particleAngle::Number=180.0,
particleAngleRange::Number=360.0,
bloomAlpha::Number=0.0,
bloomRadius::Number=6.0,
particleAlpha::Number=1.0,
tag::String="",
flag::String="",
noTexture::Bool=false,
foreground::Bool=false) =
    Entity("FemtoHelper/ParticleEmitter",
    x=x1,
    y=y1,
    particleColor=particleColor,
    particleColor2=particleColor2,
    particleTexture=particleTexture,
    particleFadeMode=particleFadeMode,
    particleRotationMode=particleRotationMode,
    particleColorMode=particleColorMode,
    particleLifespanMin=particleLifespanMin,
    particleLifespanMax=particleLifespanMax,
    particleSize=particleSize,
    particleSizeRange=particleSizeRange,
    particleScaleOut=particleScaleOut,
    particleSpeedMin=particleSpeedMin,
    particleSpeedMax=particleSpeedMax,
    particleFriction=particleFriction,
    particleAccelX=particleAccelX,
    particleAccelY=particleAccelY,
    particleFlipChance=particleFlipChance,
    particleSpinSpeedMin=particleSpinSpeedMin,
    particleSpinSpeedMax=particleSpinSpeedMax,
    particleCount=particleCount,
    particleSpawnSpread=particleSpawnSpread,
    spawnInterval=spawnInterval,
    spawnChance=spawnChance,
    particleAngle=particleAngle,
    particleAngleRange=particleAngleRange,
    bloomAlpha=bloomAlpha,
    bloomRadius=bloomRadius,
    particleAlpha=particleAlpha,
    tag=tag,
    flag=flag,
    noTexture=noTexture,
    foreground=foreground)
    Ahorn.editingOptions(entity::ParticleEmitter) = Dict{String, Any}(
  "particleColorMode" => Dict{String, Int}(
    "Static" => 0,
    "Choose" => 1,
    "Blink" => 2,
    "Fade" => 3
  ),
  "particleFadeMode" => Dict{String, Int}(
    "None" => 0,
    "Linear" => 1,
    "Late" => 2,
    "InAndOut" => 3
  ),
  "particleRotationMode" => Dict{String, Int}(
    "None" => 0,
    "Random" => 1,
    "SameAsDirection" => 2
  )
)


const placements = Ahorn.PlacementDict(
   "Particle Emitter (Femto Helper)" => Ahorn.EntityPlacement(
      ParticleEmitter
   )
)

defaultTexture = "particles/stars/00.png"

function Ahorn.selection(entity::ParticleEmitter)
    x, y = Ahorn.position(entity)

    texture = get(entity.data, "particleTexture", defaultTexture)

    return Ahorn.getSpriteRectangle(texture, x, y)
end


function getColor(color)
  if haskey(Ahorn.XNAColors.colors, color)
      return Ahorn.XNAColors.colors[color]

  else
      try
          return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

      catch

      end
  end

  return (1.0, 1.0, 1.0, 1.0)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::ParticleEmitter, room::Maple.Room)
  texture = get(entity.data, "particleTexture", defaultTexture)
  rawColor = get(entity.data, "particleColor", (1.0,1.0,1.0,1.0))
  color = getColor(rawColor)
  rawColor2 = get(entity.data, "particleColor2", (1.0,1.0,1.0,1.0))
  color2 = getColor(rawColor2)
  alpha = get(entity.data, "particlealpha", 1.0)
  boxsize = get(entity.data, "particleSpawnSpread", 8)
  Ahorn.drawSprite(ctx, texture, 0, 0, tint=color .* (1.0,1.0,1.0,alpha))
  Ahorn.drawRectangle(ctx, -boxsize ./ 2, -boxsize ./ 2, boxsize, boxsize, color2 .* (1.0, 1.0, 1.0, 0.4), color2)
end 

end

File: Loenn/entities/ParticleEmitter.lua
local drawableSprite = require("structs.drawable_sprite")
local drawing = require("utils.drawing")
local utils = require("utils")
local drawableRectangle = require("structs.drawable_rectangle")

local drawableText = require("structs.drawable_text")

local function split(inputstr, sep)
  if sep == nil then
    sep = "%s"
  end
  local t = {}
  for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
    table.insert(t, str)
  end
  return t
end

local FemtoHelperParticleEmitter = {}

FemtoHelperParticleEmitter.name = "FemtoHelper/ParticleEmitter"
FemtoHelperParticleEmitter.depth = -999999
-- FemtoHelperParticleEmitter.texture = "ahorn/ParticleEmitter"
FemtoHelperParticleEmitter.justification = {0.5, 0.5}

FemtoHelperParticleEmitter.fieldInformation = {
    particleColorMode = {
        options = {
            {"Color 1", 0},
            {"Choose Color 1/2", 1},
            {"Blink Between Color 1/2", 2},
            {"Fade Color 1 -> 2", 3}
        },
        editable = false
    },
    particleFadeMode = {
        options = {
            {"No Fade", 0},
            {"Linear Fade", 1},
            {"Late Fade", 2},
            {"In And Out", 3}
        },
        editable = false
    },
    particleRotationMode = {
        options = {
            {"Static", 0},
            {"Random", 1},
            {"Follow Direction", 2}
        },
        editable = false
    },
    particleColor = {
      fieldType = "color",
      useAlpha = true
    },
    particleColor2 = {
      fieldType = "color",
      useAlpha = true
    }
}
FemtoHelperParticleEmitter.placements = {
    name = "default",
    data = {
      particleColor="FF7777FF",
      particleColor2="9C2E36FF",
      particleTexture="particles/feather",
      particleFadeMode=2,
      particleColorMode=3,
      particleRotationMode=1,
      particleLifespanMin=1.25,
      particleLifespanMax=1.75,
      particleSize=1.5,
      particleSizeRange=0.5,
      particleScaleOut=true,
      particleSpeedMin=10.0,
      particleSpeedMax=20.0,
      particleFriction=50.0,
      particleAccelX=30.0,
      particleAccelY=-60.0,
      particleFlipChance=true,
      particleSpinSpeedMin=4.0,
      particleSpinSpeedMax=8.0,
      particleCount=1,
      particleSpawnSpread=4,
      spawnInterval=0.1,
      spawnChance=90.0,
      particleAngle=180.0,
      particleAngleRange=360.0,
      bloomAlpha=0.0,
      bloomRadius=6.0,
      particleAlpha=1.0,
      foreground=false,
      tag="",
      flag="",
      noTexture = false,
      attachToPlayer = false,
      attachToPlayerOffsetX = 0,
      attachToPlayerOffsetY = 0,
    }
}

local function getColor(color)
    local success, r, g, b = utils.parseHexColor(color)
    return success and {r, g, b} or {1, 1, 1}
end

function FemtoHelperParticleEmitter.sprite(room, entity)
  local emitter_texture = "loenn/FemtoHelper/ParticleEmitter"
  local emitter_textureOutline = "loenn/FemtoHelper/ParticleEmitterOutline"
  local particle_texture = entity.particleTexture
  
  local color1 = getColor(entity.particleColor)
  local color2 = getColor(entity.particleColor2)

  local MainSprite = drawableSprite.fromTexture(emitter_texture, entity)
  local MainSpriteOutline = drawableSprite.fromTexture(emitter_textureOutline, entity)

  MainSprite:setColor(color1)
  MainSpriteOutline:setColor(color2)

  local Particle1 = drawableSprite.fromTexture(particle_texture, entity)
  local Particle2 = drawableSprite.fromTexture(particle_texture, entity)

  if Particle1 and Particle2 then
    Particle1:addPosition(-8, -8)
    Particle2:addPosition(8, -8)

    Particle1:setColor(color1)
    Particle2:setColor(color2)

    Particle1:setScale(entity.particleSize - entity.particleSizeRange)
    Particle2:setScale(entity.particleSize + entity.particleSizeRange)
  end

  return {
    drawableRectangle.fromRectangle("bordered", entity.x - entity.particleSpawnSpread, entity.y - entity.particleSpawnSpread, entity.particleSpawnSpread * 2, entity.particleSpawnSpread * 2, {color2[1], color2[2], color2[3], 0.25}, {color1[1], color1[2], color1[3], 0.25}),
    MainSprite,
    MainSpriteOutline,
    Particle1,
    Particle2,
  }
end

function FemtoHelperParticleEmitter.selection(room, entity)
    local x, y = entity.x, entity.y

    local mainRectangle = utils.rectangle(x-6, y-6, 12, 12)
    return mainRectangle
end

return FemtoHelperParticleEmitter